lib/profiling/nvidia/nvtx_prof.f90

```diff
@@ -114,51 +117,52 @@ end subroutine ProfileInit
   !> Starts a profiling area. The module and region name can be used to create
   !! a unique name for each region.
   !! Parameters: 
+  !! this:         This PSyData instance - holds the selected colour and
+  !!               name of this region.
   !! module_name:  Name of the module in which the region is
   !! region_name:  Name of the region (could be name of an invoke, or
   !!               subroutine name).
-  !! profile_data: Persistent data - holds the selected colour and name of
-  !!               this region.
-  subroutine ProfileStart(module_name, region_name, profile_data)
+  !! this: Persistent data 
+  subroutine PreStart(this, module_name, region_name)
     implicit none
+    class(PSyDataType), intent(inout) :: this
     character*(*), intent(in) :: module_name, region_name
-    type(ProfileData), target, intent(inout) :: profile_data
     ! Locals
     type(nvtxEventAttributes) :: event
 
-    if(.not. profile_data%initialised)then
+    if(.not. this%initialised)then
        ! This is the first time we've seen this region. Construct and
        ! save its name to save on future string operations.
-       profile_data%initialised = .true.
+       this%initialised = .true.
 
        ! Round-robin the colour of each region created.
        if (last_colour < NUM_COLOURS) then
           last_colour = last_colour + 1
        else
           last_colour = 1
        end if
-       profile_data%colour_index = last_colour
+       this%colour_index = last_colour
 
-       profile_data%name = trim(module_name)//":"//trim(region_name) &
+       this%name = trim(module_name)//":"//trim(region_name) &
                            &   //C_NULL_CHAR
     end if
     
-    event%color = col(profile_data%colour_index)
-    event%message = c_loc(profile_data%name)
+    event%color = col(this%colour_index)
+    event%message = c_loc(this%name)
```

Comment:
Now this opens a whole can of worms: your original code is actually incorrect:: if you have a dummy parameter that has the target attribute, then the actual parameter must also have the target attribute (http://www.cs.rpi.edu/~szymansk/OOF90/bugs.html#6, see explanation at very end). But ProfileData (then, now PSyData) is not declared with target attribute, so it is not correct to pass this to a subroutine that assumes target.
I don't think this will actually cause a problem (since the ProfileData/PSyData objects are static, so they should stay around in memory), I am not sure what the best way is to proceed: we either add target to the declaration of the static objects, or remove target in nvtx_prof (then we would have to allocate memory and duplicate the string I guess??)

I have fixed the compilation error, so the code is as good as it was before. I would suggest we open this as a ticket for further API improvements - just to stream-line this PR. It's quick for me to fix if you want to change the API to include target.

Reply:
It compiles now and I'm happy for you to open another ( :-( ) issue for the 'target' problem.

Reply:
Please reference the new issue from this code.